core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/HttpChecksumStage.java

```diff
@@ -0,0 +1,273 @@
+            return request;
+        }
+
+        if (flexibleChecksumInHeaderRequired(context, resolvedChecksumSpecs)) {
+            addFlexibleChecksumInHeader(request, context, resolvedChecksumSpecs);
+            return request;
+        }
+
+        return request;
+    }
+
+    private boolean md5ChecksumRequired(SdkHttpFullRequest.Builder request, RequestExecutionContext context) {
+        boolean isHttpChecksumRequired =
+            context.executionAttributes().getAttribute(SdkInternalExecutionAttribute.HTTP_CHECKSUM_REQUIRED) != null ||
+            HttpChecksumUtils.isMd5ChecksumRequired(context.executionAttributes());
+
+        boolean requestAlreadyHasMd5 = !request.firstMatchingHeader(Header.CONTENT_MD5).isPresent();
+
+        if (!isHttpChecksumRequired || requestAlreadyHasMd5) {
+            return false;
+        }
+
+        if (context.requestProvider() != null) {
+            throw new IllegalArgumentException("This operation requires a content-MD5 checksum, but one cannot be calculated "
+                                               + "for non-blocking content.");
+        }
+
+        return context.executionContext().interceptorContext().requestBody().isPresent();
+    }
+
+    /**
+     * Implements the "httpChecksumRequired" C2J trait. Operations with that trait applied will automatically include a
+     * "Content-MD5" header, containing a checksum of the payload.
+     *
+     * <p>This is NOT supported for asynchronous HTTP content, which is currently only used for streaming upload operations.
+     * If such operations are added in the future, we'll have to find a way to support them in a non-blocking manner. That will
+     * likely require interface changes of some sort, because it's not currently possible to do a non-blocking update to
+     * request headers.
+     *
+     * <p>
+     * Calculates the MD5 checksum of the provided request (and base64 encodes it), and adds the header to the request.
+     *
+     * <p>Note: This assumes that the content stream provider can create multiple new streams. If it only supports one (e.g. with
+     * an input stream that doesn't support mark/reset), we could consider buffering the content in memory here and updating the
+     * request body to use that buffered content. We obviously don't want to do that for giant streams, so we haven't opted to do
+     * that yet.
+     */
+    private void addMd5ChecksumInHeader(SdkHttpFullRequest.Builder request, RequestExecutionContext context) {
+        try {
+            String payloadMd5 = Md5Utils.md5AsBase64(request.contentStreamProvider().newStream());
+            request.putHeader(Header.CONTENT_MD5, payloadMd5);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private boolean flexibleChecksumInTrailerRequired(RequestExecutionContext context, ChecksumSpecs checksumSpecs) {
+        boolean hasRequestBody = true;
```

Comment:
before there were separate interceptors so it was initialized by checking if the sync/async requestbody is present. Here I had to initialize to value first, but the ClientType always be either SYNC or ASYNC so this will get overwritten